generator/main.cpp

```diff
@@ -132,36 +137,205 @@ int main(int argc, char *argv[])
+         * the MacOS file system layout.
+         *
+         * TODO: remove this, require the directories to be specified on the
+         * command line.
+         */
+        QString qtdir = getenv ("QTDIR");
+
+        if (qtdir.isEmpty()) {
+#if defined(Q_OS_MAC)
+            qWarning(
+"QTDIR environment variable not set. Assuming standard binary install using\n"
+"frameworks.");
+            foreach (const QString &mod, coreModules)
+                includes << ("/Library/Frameworks/Qt" + mod + ".framework/Headers");
+            includes << "/Library/Frameworks"; // this seems wrong
+#else
+            qWarning(
+"QTDIR environment variable not set. This may cause problems with finding the\n"
+"necessary include files.  You can find the correct directory with the\n"
+"command:\n\n"
+"  qmake -query QT_INSTALL_HEADERS\n\n"
+"This directory can be supplied using the command line argument --qt-headers\n"
+            );
+#endif
+        } else {
+            /* Legacy handling: Qt install into a single directory on its own,
+             * basically /opt/qt or something like that:
+             */
+            qtHeaders = qtdir + "/include";
+        }
+    }
+
+    if (!qtHeaders.isEmpty()) {
+        /* Look for the 'standard' directories, favouring the first found
+         * in the qt-headers list.
+         *
+         * NOTE: use of a list here is intended for the project build workflows
+         * which apparently have to assemble the headers from difference
+         * installation directories.  This could be avoided by using
+         * --include-paths instead!
+         *
+         */
+        QStringList dirList;
+        QStringList requiredModules(coreModules);
+
+        foreach (const QString &dir, qtHeaders.split(path_splitter))
+            if (QDir(dir).exists())  {
+                QStringList remaining(requiredModules);
+
+                foreach (const QString &mod, requiredModules) {
+                    const QString modpath(dir + "/Qt" + mod);
+                    if (QDir(modpath).exists()) {
+                        includes << modpath;
+                        remaining.removeAll(mod);
+                    }
+                }
+
+                requiredModules = remaining;
+                dirList << dir;
+            } else
+                qWarning((dir + " Qt header directory not found")
+                                            .toLocal8Bit().constData());
+
+        /* Add the top level directories at the end; this simulates the legacy
+         * behaviour.
+         */
+        if (!minimal)
+            includes += dirList;
+
+        /* Warn if standard directories are not found: */
+        if (!requiredModules.isEmpty()) {
+            QString sep("");
+            QString errMsg("WARNING: missing core Qt modules:");
+
+            foreach (const QString &mod, requiredModules) {
+                errMsg += sep + " Qt" + mod;
```

Comment:
Probably easier to add the "Qt" right at the beginning and use `join(", ")` here...

Reply:
> the baroque nature of the rest of the code

I like that 😁